Portaudio 19.7.0 fixes: invalid device error and mingw gcc build

dimi309 opened the pull request:
Hi everyone,

The packaged portaudio library (version 19.7.0) was not working with my programs on Linux (tried on Debian, Ubuntu and Fedora). Pa_OpenStream would fail to open a stream, reporting an invalid device error. I have noticed that the difference of the package from a simple portaudio cmake build (which works for me without issue) was that the package's conanfile.py was always passing PA_USE_JACK and PA_USE_ALSA definitions to cmake and, on top of that, those were set to true by default. But the code in portaudio's CMakeLists.txt file (the one that comes with the original library)  can detect support for alsa and jack and "decide" on its own whether to use them, when no definitions are passed, for example: 
```
FIND_PACKAGE(Jack)
    IF(JACK_FOUND)
      OPTION(PA_USE_JACK "Enable support for Jack" ON)
    ELSE()
      OPTION(PA_USE_JACK "Enable support for Jack" OFF)
ENDIF()
```
The only way to get the package to work with my programs on Linux was to re-enable this possibility, by only passing PA_USE_JACK and PA_USE_ALSA if the corresponding package options are set to true:
```
 if self.options.get_safe("with_jack", False):
    self._cmake.definitions["PA_USE_JACK"] = True
 if self.options.get_safe("with_alsa", False):
    self._cmake.definitions["PA_USE_ALSA"] = True
```
So now portaudio has stopped producing the invalid device error on my machines, while the package still provides the possibility to enable jack for example, if a developer so chooses.

I am also linking to setupapi for windows gcc builds (mingw), because I have had some undefined symbols there.

Finally, I am also linking to and deploying the alsa library. On actual linux machines it is usually there already, but it may be missing from some online CI environment images. It was not deployed by default on an appveyor image I am using for example. 

I don't know if you find these changes acceptable, but I thought to at least share them with you. I have tested them on the three linux distributions I have mentioned above and they have resolved the issues I have had. Also, here are some builds of a little game I have put together, which uses this new version of the portaudio package: https://ci.appveyor.com/project/dimi309/gloom-game-conan/builds/40999713 
You can also see the mingw build working well there.

dimi309 commented:
I have some additional information. I have debugged the initialisation of portaudio extensively in my code and I can now see more of what is happening. For whatever reason that I do not understand, when use_alsa or use_jack are explicitly passed via the conanfile to cmake, Pa_GetDeviceCount and Pa_GetDeviceInfo do not find any available devices and so the default one returned is thus, as reported by an error, invalid. When left up to portaudio's CMakeLists (meaning that the conanfile does not even pass false for use_alsa or use_jack), some 19 ALSA devices are detected, including a default one. This happens regardless whether jack is installed or not. Either way Portaudio never detects jack, in my code at least, not even as an available API.
As I have mentioned, the code in my pull request allows us (or me at least), to have a functional portaudio build, by allowing us to not even tell cmake whether we want alsa, jack or both (use_alsa = False in conanfile => no CMAKE variable, not even false). And the option to pass True remains available to whoever wishes to experiment with the problem I have reported.

dimi309 commented:
I have finally figured it out. The source of the problem is the libalsa package at conancenter. With my changes it is made possible to not pass a PA_WITH_ALSA parameter to cmake when building portaudio. But this does not prevent portaudio from using alsa at all. It detects the alsa library deployed with the linux distribution and uses that one and this is why it works in this case. When with_alsa is set to True in the conanfile, then not only is PA_WITH_ALSA=ON  sent to the portaudio build, but also libalsa/1.1.9 from conancenter is used. And then, no sound devices are made available in the program. I have tried to upgrade the portaudio package to use libalsa/1.2.5.1 but the problem persists. It only works when portaudio is made to use the linux-distributed alsa library. 

The fact remains that my changes do help sidestep the issue. Setting with_alsa = False will send no PA_WITH_ALSA parameter to the portaudio build, which will make cmake detect the system's alsa library and use it, thus working properly. At the same time, setting with_alsa to TRUE still use conancenter's libalsa, in case the current or a future version works on some machines. But I understand if you would rather not merge my pull request until conancenter's libalsa starts working. Of course I will note at this point that in the case of jack, the portaudio package is using the linux distro's library in all cases so, in my humble opinion, there is no contradiction in using the linux distro alsa library too.

dimi309 commented:
So I have restored with_jack to work as it originally did. The issue was only caused by with_alsa. So, to sum it all up, this pull request's code does the following:

- When with_alsa is set to False alsa is still used, just not the conancenter version but rather the linux distro's library
- When with_alsa is set to True, concanenter's alsa is used (good luck with that :worried: )
- with_jack contintues to work as it did
- Windows gcc (mingw) builds link to the setupapi library, thus avoiding undefined symbol errors

dimi309 commented:
I have mentioned the problem in an issue on conan-center-index too: https://github.com/conan-io/conan-center-index/issues/7526

dimi309 commented:
Since this pull request has not been picked up for a while, I am closing it.

Croydon commented:
I'm sorry that you never got an answer to this pull request.

For me, there are many unanswered questions around audio on Linux and how we should handle it in Conan.

This PR also changed default values and introduced more system_packages which we generally don't like, but if it is broken currently, then it obviously needs to get fixed.

Croydon commented on recipes/portaudio/19.7.0/conanfile.py:
```diff
@@ -55,10 +55,12 @@ def system_requirements(self):
         if self.settings.os == "Linux":
             if tools.os_info.with_apt:
                 installer = tools.SystemPackageTool()
+                installer.install("libasound2-dev") 
```
What is the relationship between the `libalsa/*` Conan package and those system_packages?

Is the Conan package broken? Do we need both?

dimi309 replied:
When with_alsa is set to False the linux distro's library is used, which is why I am installing it. Perhaps I should have added a condition here to not install the linux package if with_alsa is set to true, Because when with_alsa is set to true, the conancenter's alsa is used. This was my hacky way of allowing consumers of the package to use conancenter's alsa if they chose to, even though it never worked for me.

Croydon commented on recipes/portaudio/19.7.0/conanfile.py:
```diff
@@ -74,7 +76,10 @@ def _configure_cmake(self):
-            self._cmake.definitions["PA_USE_ALSA"] = self.options.get_safe("with_alsa", False)
+            if self.options.get_safe("with_alsa", False): # with_alsa=False just makes portaudio use the Linux distro's alsa
+                                                          # as a workaround to the fact that conancenter's alsa does not work,
+                                                          # at least some of the time (no devices detected by portaudio)
```
Does this mean, that it works sometimes? If yes, under what circumstances?

dimi309 replied:
Conancenter's alsa has never worked for me, but since it was referenced by this packaged I supposed it had been tested and it had worked for somebody else so I did not want my comment to be too absolute.

dimi309 commented:
> I'm sorry that you never got an answer to this pull request.
> 
> For me, there are many unanswered questions around audio on Linux and how we should handle it in Conan.
> 
> This PR also changed default values and introduced more system_packages which we generally don't like, but if it is broken currently, then it obviously needs to get fixed.

Thanks for your message @Croydon ! No worries, I don't expect everything to be answered on github. Sometimes there is not enough time to follow up on every published open source bit. I wasn't bitter or anything, just stating the fact 😄
 
I understand your point about not using system packages and, normally I tend not to either. I have been using conan on linux for a long time however and, sometimes, I have found it to be the only realistic solution; especially when the package is used to handle I/O or sound at such a low level that it is almost a driver. Alsa would fall under this category in my humble opinion. It might be a bit of a challenge making a reliable conan package out of it. I had actually tried forking and fixing the alsa conan package myself but I couldn't get it to work. So I thought to provide this "escape" option in the portaudio package. But I understand if you find this too "hacky" 😄
